module/zfs/vdev_mirror.c

```diff
@@ -637,8 +638,15 @@ vdev_mirror_io_start(zio_t *zio)
 	}
 
 	if (zio->io_type == ZIO_TYPE_READ) {
-		if (zio->io_bp != NULL &&
-		    (zio->io_flags & ZIO_FLAG_SCRUB) && !mm->mm_resilvering) {
+		boolean_t have_checksum = zio->io_bp != NULL;
+		boolean_t is_scrub = have_checksum &&
+		    (zio->io_flags & ZIO_FLAG_SCRUB) &&
+		    !mm->mm_resilvering;
+		boolean_t resilver_read = have_checksum &&
+		    (zio->io_flags & (ZIO_FLAG_SCRUB | ZIO_FLAG_RESILVER)) &&
+		    (zio_checksum_table[BP_GET_CHECKSUM(zio->io_bp)].ci_flags &
```

Comment:
This would be a good place to add a comment explaining the reasoning behind this.